src/chainstate/coordinator/tests.rs

```diff
@@ -2287,6 +2289,319 @@ fn test_stx_transfer_btc_ops() {
     }
 }
 
+
+#[test]
+fn test_delegate_stx_btc_ops() {
+    let path = "/tmp/stacks-blockchain-delegate-stx-btc-ops";
+    let _r = std::fs::remove_dir_all(path);
+
+    let pox_v1_unlock_ht = 12;
+    let pox_consts = Some(PoxConstants::new(5, 3, 3, 25, 5, pox_v1_unlock_ht));
+    let burnchain_conf = get_burnchain(path, pox_consts.clone());
+
+    let vrf_keys: Vec<_> = (0..50).map(|_| VRFPrivateKey::new()).collect();
+    let committers: Vec<_> = (0..50).map(|_| StacksPrivateKey::new()).collect();
+
+    let first_del = p2pkh_from(&StacksPrivateKey::new());
+    let second_del = p2pkh_from(&StacksPrivateKey::new());
+    let rewards = pox_addr_from(&StacksPrivateKey::new());
+    let recipient_addr = p2pkh_from(&StacksPrivateKey::new());
+    let balance = 6_000_000_000 * (core::MICROSTACKS_PER_STACKS as u64);
+    let delegated_amt = 1_000_000_000 * (core::MICROSTACKS_PER_STACKS as u128);
+    let initial_balances = vec![
+            (first_del.clone().into(), balance),
+            (second_del.clone().into(), balance)
+        ];
+
+    setup_states(
+        &[path],
+        &vrf_keys,
+        &committers,
+        pox_consts.clone(),
+        Some(initial_balances),
+        StacksEpochId::Epoch21,
+    );
+
+    let mut coord = make_coordinator(path, Some(burnchain_conf.clone()));
+
+    coord.handle_new_burnchain_block().unwrap();
+
+    let sort_db = get_sortition_db(path, pox_consts.clone());
+
+    let tip = SortitionDB::get_canonical_burn_chain_tip(sort_db.conn()).unwrap();
+    assert_eq!(tip.block_height, 1);
+    assert_eq!(tip.sortition, false);
+    let (_, ops) = sort_db
+        .get_sortition_result(&tip.sortition_id)
+        .unwrap()
+        .unwrap();
+
+    // we should have all the VRF registrations accepted
+    assert_eq!(ops.accepted_ops.len(), vrf_keys.len());
+    assert_eq!(ops.consumed_leader_keys.len(), 0);
+
+    let mut started_first_reward_cycle = false;
+    // process sequential blocks, and their sortitions...
+    let mut stacks_blocks: Vec<(SortitionId, StacksBlock)> = vec![];
+    let mut anchor_blocks = vec![];
+
+    // track the reward set consumption
+    // let mut reward_cycle_count = 0;
+    // let mut reward_recipients = HashSet::new();
+    for ix in 0..vrf_keys.len() {
+        info!("ix: {:?}", ix); 
+        let vrf_key = &vrf_keys[ix];
+        let miner = &committers[ix];
+
+        let mut burnchain = get_burnchain_db(path, pox_consts.clone());
+        let mut chainstate = get_chainstate(path);
+
+        // The chain will look something like this 
+        // B0 -> B1 -> B2 -> ... B20 -> B21 
+        //                          \ _ B22
+        //                          \ _ B23
+        //                          \ _ B24
+        //                            ....
+        //                          \ _ B30 -> B31 -> ...
+        let parent = if ix == 0 {
+            BlockHeaderHash([0; 32])
+        } else if ix >= 22 && ix <= 30 {
```

Comment:
I think you want `ix >= 21` here.  Otherwise, with `ix >= 22`, blocks 22 through 30 all build off of `ix == 21`, in which the delegation for `delegate_amt * 2` will have taken effect.

Reply:
The `delegate-stx` op materializes in `ix` blocks 21 through 27, because you send it at `ix == 20`.  So, blocks 21 through 27 must all be siblings in order to test that the burnchain operations can manifest in all Stacks blocks mined between n+1 and n+7 burnchain blocks after the burnchain operation confirms in burnchain block n.

Reply:
The weirdness in the indexing here is caused by the fact that the logic for the check is happening before the burnchain/stacks block is being handled. 

The logic takes the following structure, which makes it difficult to follow: 
- Create burnchain ops 
- Check the delegation amount state
- Handle the burnchain and stacks block 

ix = 20, the delegate stx burnchain op created and handled. The op is sent in the burnchain block created in this iteration.
In ix = 21, the check for delegation state occurs before the stacks block is handled. That's why I do the check in the next iteration (which I agree is confusing). I'm moving the delegation state check to after the stacks block is handled to make it easier to reason about.

Reply:
Update: the issue was that the stacks tip was not updating since the forks were only 1 block long. 
I had to compute the stacks tip another way for the test to work.